includes/data-stores/class-wcs-related-order-store-cached-cpt.php

```diff
@@ -750,16 +751,65 @@ public function get_related_order_ids_by_types( WC_Order $subscription, $related
 		$related_order_ids = [];
 
 		// Declare batch processing mode for this subscription.
-		self::$batch_processing_related_orders[ $subscription_id ] = true;
+		$cache_key = $this->start_batch_processing_mode( $subscription );
 
 		foreach ( $related_order_types as $relation_type ) {
 			$related_order_ids[ $relation_type ] = $this->get_related_order_ids( $subscription, $relation_type );
 		}
 
-		// Unset the batch processing mode for this subscription.
-		unset( self::$batch_processing_related_orders[ $subscription_id ] );
-		unset( self::$subscription_meta_cache[ $subscription_id ] );
+		$this->stop_batch_processing_mode( $cache_key );
 
 		return $related_order_ids;
 	}
+
+	/**
+	 * Starts batch processing mode for a subscription.
+	 *
+	 * @param WC_Subscription $subscription The subscription to start batch processing mode for.
+	 * @return string The cache key for the subscription.
+	 */
+	private function start_batch_processing_mode( $subscription ) {
+		$cache_key = $this->get_batch_processing_cache_key( $subscription );
+
+		self::$batch_processing_related_orders[ $cache_key ] = true;
+		return $cache_key;
+	}
+
+	/**
+	 * Stops batch processing mode for a subscription.
+	 *
+	 * Destroys the cache and removes the cache key.
+	 *
+	 * @param string $cache_key The batch processing cache key.
+	 */
+	private function stop_batch_processing_mode( $cache_key ) {
+		unset( self::$batch_processing_related_orders[ $cache_key ] );
+		unset( self::$subscription_meta_cache[ $cache_key ] );
+	}
+
+	/**
+	 * Checks if batch processing mode is active for a subscription.
+	 *
+	 * @param string $cache_key The batch processing cache key.
+	 * @return bool True if batch processing mode is active, false otherwise.
+	 */
+	private function is_batch_processing( $cache_key ) {
+		return isset( self::$batch_processing_related_orders[ $cache_key ] );
+	}
+
+	/**
+	 * Gets the batch processing cache key for a subscription.
+	 *
+	 * The cache key is a unique combination of the subscription ID and the data store class name.
+	 *
+	 * @param WC_Subscription $subscription The subscription to get the cache key for.
+	 * @param bool|object     $data_store   The data store which will be used to read the subscription meta. Defaults to the current subscription's data store.
+	 *
+	 * @return string The cache key for the subscription.
+	 */
+	private function get_batch_processing_cache_key( $subscription, $data_store = null ) {
+		// If no data store is provided, use the subscription object's data store or load the default data store if no subscription data store is found.
+		$data_store = $data_store ?? $subscription->get_data_store() ?? DataStore::load( 'subscriptions' );
+		return get_class( $data_store ) . '-' . $subscription->get_id();
```

Comment:
While doing some quick testing, I noticed that calling `get_class()` on $data_store always returns `WC_Data_Store`.

So it looks like we need to use `get_current_class_name()` instead.

```suggestion
		return ( is_a( $data_store, 'WC_Data_Store' ) ? $data_store->get_current_class_name() . '-' : '' ) . $subscription->get_id();
```

Reply:
Wow. That's a crazy good catch. I pushed [3fc5524](https://github.com/Automattic/woocommerce-subscriptions-core/pull/804/commits/3fc5524282a1f210b90e4cc50550a5644382435d) which is basically the same change you suggested, except I split it out to a separate line just to help with readability.